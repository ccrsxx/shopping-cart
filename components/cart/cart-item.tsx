import Link from 'next/link';
import { motion } from 'framer-motion';
import { useShoppingCart } from '@lib/context/shopping-cart';
import { formatCurrency } from '@lib/currency';
import { setTransition } from '@lib/transition';
import { MdAdd, MdRemove, MdDelete } from '@assets/icons';
import { ImageLoader } from '@components/ui/image-loader';
import { Button } from '@components/ui/button';

type CartItemProps = {
  id: number;
  index: number;
  title: string;
  image: string;
  price: number;
  quantity: number;
  toggleCart: () => void;
};

export function CartItem({
  id,
  index,
  title,
  image,
  price,
  quantity,
  toggleCart
}: CartItemProps): JSX.Element {
  const { deleteProduct, handleProductQuantity } = useShoppingCart();

  const toProduct = `/product/${id}`;

  return (
    <motion.li
      {...setTransition({
        layout: true,
        delayIn: index * 0.02,
        direction: 'right',
        durationOut: 0.25
      })}
    >
      <div className='flex rounded-lg border border-border-primary'>
        <Link href={toProduct}>
          <a
            className='tab hidden rounded-r-none transition hover:brightness-90 
                       focus-visible:brightness-90 md:block'
            onClick={toggleCart}
          >
            <ImageLoader
              divStyle='h-[108px] w-[108px] rounded-l-lg shrink-0 shrink-0 bg-white'
              imageStyle='!p-4'
              src={image}
              alt={title}
            />
          </a>
        </Link>
        <div className='flex w-full flex-col justify-between py-2 px-4'>
          <div>
            <Link href={toProduct}>
              <a
                className='tab overflow-hidden text-ellipsis font-medium text-white/90
                           transition [display:-webkit-box] [-webkit-line-clamp:1]
                           [-webkit-box-orient:vertical] hover:text-white'
                onClick={toggleCart}
              >
                {title}
              </a>
            </Link>
            <p className='font-bold'>{formatCurrency(price)}</p>
          </div>
          <div className='flex flex-wrap justify-between text-sm font-light'>
            <div className='flex gap-1'>
              <p>Quantity: </p>
              <input
                className='max-w-[60px] rounded-lg bg-background px-1 transition hover:ring-2 
                           hover:ring-accent focus:outline-none focus:ring-2 focus:ring-accent'
                type='number'
                min={1}
                max={10_000}
                value={quantity}
                onChange={handleProductQuantity(id)}
              />
            </div>
            <p>
              Total:{' '}
              <span className='font-semibold'>
                {formatCurrency(price * quantity)}
              </span>
            </p>
          </div>
        </div>
        <div className='grid border-l border-border-primary text-sm inner:rounded-none'>
          <Button
            className='!rounded-tr-lg disabled:text-gray-600 disabled:!brightness-100'
            ariaLabel='Increase'
            Icon={MdAdd}
            onClick={handleProductQuantity(id, 'increment')}
            disabled={quantity >= 10_000}
          />
          <Button
            className='border-y border-border-primary disabled:text-gray-600 disabled:!brightness-100'
            ariaLabel='Decrease'
            Icon={MdRemove}
            onClick={handleProductQuantity(id, 'decrement')}
            disabled={quantity <= 1}
          />
          <Button
            className='!rounded-br-lg'
            ariaLabel='Delete'
            Icon={MdDelete}
            onClick={deleteProduct(id)}
          />
        </div>
      </div>
    </motion.li>
  );
}
