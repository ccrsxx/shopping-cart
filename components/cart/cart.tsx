import { useState, useEffect } from 'react';
import Link from 'next/link';
import { AnimatePresence, motion } from 'framer-motion';
import { MdShoppingCart, MdArrowForward } from '@assets/icons';
import { Button } from '@components/ui/button';
import { useShoppingCart } from '@lib/context/shopping-cart';
import { formatCurrency } from '@lib/currency';
import { setTransition } from '@lib/transition';
import { CartItem } from './cart-item';

export function Cart(): JSX.Element {
  const { cartProducts, currentCart, totalPrice, isMobile, clearCart } =
    useShoppingCart();

  const [totalProducts, setTotalProducts] = useState(0);
  const [isCartOpen, setIsCartOpen] = useState(false);

  useEffect(() => {
    setTotalProducts(cartProducts);
  }, [cartProducts]);

  useEffect(() => {
    if (isMobile && !isCartOpen) document.documentElement.style.overflow = '';

    document[isMobile ? 'documentElement' : 'body'].style.overflowY = isCartOpen
      ? 'hidden'
      : '';
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isCartOpen]);

  const toggleCart = (): void => setIsCartOpen(!isCartOpen);

  const cartLength = currentCart.length;

  return (
    <>
      <Button className='relative !p-2' ariaLabel='Open' onClick={toggleCart}>
        <MdShoppingCart size={24} />
        <span
          className='absolute -right-0.5 -top-0.5 flex h-5 min-w-[20px] items-center
                     justify-center rounded-full bg-main-red p-1 text-xs'
        >
          {totalProducts}
        </span>
      </Button>
      <AnimatePresence exitBeforeEnter>
        {isCartOpen && (
          <div className='fixed inset-0 z-10'>
            <motion.div
              className='fixed h-screen w-screen bg-black/80'
              onClick={toggleCart}
              {...setTransition({ direction: 'none' })}
            />
            <motion.div
              className='fixed top-0 -right-4 h-screen w-[500px] max-w-[90vw]
                         rounded-l-lg bg-background p-8 pr-12'
              {...setTransition({
                direction: 'right',
                distance: 'full',
                durationIn: 0.5
              })}
            >
              <div className='flex h-full flex-col gap-4'>
                <div className='flex justify-between'>
                  <h2 className='text-2xl font-bold'>
                    {cartLength
                      ? `${cartLength} product${cartLength > 1 ? 's' : ''}`
                      : 'No product added'}
                  </h2>
                  <Button
                    className='!justify-end !p-0 text-xl normal-case text-secondary hover:scale-100 hover:!bg-inherit'
                    label='Clear'
                    onClick={clearCart}
                  />
                </div>
                <ul
                  className='relative order-1 -mx-8 -my-1 -mr-6 flex flex-1 flex-col gap-4
                             overflow-y-auto overflow-x-hidden py-1 px-8 pr-6'
                >
                  <AnimatePresence>
                    {currentCart.map((cartProduct, index) => (
                      <CartItem
                        index={index}
                        toggleCart={toggleCart}
                        key={cartProduct.id}
                        {...cartProduct}
                      />
                    ))}
                  </AnimatePresence>
                </ul>
                <div className='flex flex-wrap items-center justify-between md:order-2'>
                  <p className='text-lg'>
                    Total:{' '}
                    <span className='font-bold'>
                      {formatCurrency(totalPrice)}
                    </span>
                  </p>
                  <Link href='/checkout'>
                    <a tabIndex={-1} onClick={toggleCart}>
                      <Button
                        Icon={MdArrowForward}
                        label='Checkout'
                        className='text-xl normal-case hover:!bg-inherit hover:!text-accent'
                        flipped
                      />
                    </a>
                  </Link>
                </div>
              </div>
            </motion.div>
          </div>
        )}
      </AnimatePresence>
    </>
  );
}
