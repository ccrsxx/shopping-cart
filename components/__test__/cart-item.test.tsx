/* eslint-disable @typescript-eslint/no-explicit-any */

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { CartItem } from '@components/cart/cart-item';

interface MockProps {
  deleteProduct: (productId: number) => () => void;
  handleProductQuantity: (
    productId: number,
    type?: 'increment' | 'decrement'
  ) => (e?: React.ChangeEvent<HTMLInputElement>) => void;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const mockDeleteProduct = jest.fn((productId: number) => (): void => undefined);

const currentCart = {
  id: 24,
  index: 1,
  title: 'Emilia',
  image: '/images/emilia.jpg',
  price: 10,
  quantity: 1,
  toggleCart: (): void => undefined
};

const mockHandleProductQuantity =
  (productId: number, type?: 'increment' | 'decrement') =>
  (e?: React.ChangeEvent<HTMLInputElement>): void => {
    let inputValue = !type ? parseInt(e?.target.value as string, 10) : null;

    if (!type && !inputValue) inputValue = 1;
    else if (inputValue && inputValue >= 10_000) inputValue = 10_000;

    currentCart.quantity =
      inputValue ??
      (type === 'increment'
        ? currentCart.quantity + 1
        : currentCart.quantity - 1);
  };

jest.mock('@lib/context/shopping-cart', () => ({
  useShoppingCart: (): MockProps => ({
    deleteProduct: mockDeleteProduct,
    handleProductQuantity: mockHandleProductQuantity
  })
}));

let rerenderComponent: (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>
) => void;

describe('test CartItem', () => {
  beforeEach(() => {
    const { rerender } = render(<CartItem {...currentCart} />);
    rerenderComponent = rerender;
  });

  it('renders correctly', () => {
    const { container } = render(<CartItem {...currentCart} />);
    expect(container).toMatchSnapshot();
  });

  it('should call deleteProduct when delete button is clicked', () => {
    const deleteButton = screen.getByRole('button', { name: /delete/i });

    userEvent.click(deleteButton);

    expect(mockDeleteProduct).toHaveBeenCalledWith(24);
  });

  it('shows the current quantity on the input', () => {
    const input = screen.getByRole('spinbutton');

    rerenderComponent(<CartItem {...currentCart} />);

    expect(input).toHaveValue(1);
  });

  it('should call handleProductQuantity when input is changed', () => {
    const input: HTMLInputElement = screen.getByRole('spinbutton');

    userEvent.type(input, '200');

    rerenderComponent(<CartItem {...currentCart} />);

    expect(input).toHaveValue(1200);
  });

  it('should not surpass the minimum and maximum quantity with input', () => {
    const input: HTMLInputElement = screen.getByRole('spinbutton');

    userEvent.type(input, '-1');

    rerenderComponent(<CartItem {...currentCart} />);

    expect(input).toHaveValue(1);

    userEvent.type(input, '20_000');

    rerenderComponent(<CartItem {...currentCart} />);

    expect(input).toHaveValue(10_000);
  });

  it('should not surpass the minimum and maximum quantity with button', () => {
    currentCart.quantity = 1;

    rerenderComponent(<CartItem {...currentCart} />);

    const input = screen.getByRole('spinbutton');
    const incrementButton = screen.getByRole('button', { name: /increase/i });
    const decrementButton = screen.getByRole('button', { name: /decrease/i });

    userEvent.click(decrementButton);

    expect(decrementButton).toBeDisabled();
    expect(currentCart.quantity).toBe(1);

    userEvent.click(incrementButton);
    userEvent.click(incrementButton);

    rerenderComponent(<CartItem {...currentCart} />);

    expect(input).toHaveValue(3);

    currentCart.quantity = 10_000;

    rerenderComponent(<CartItem {...currentCart} />);

    userEvent.click(incrementButton);

    rerenderComponent(<CartItem {...currentCart} />);

    expect(incrementButton).toBeDisabled();
    expect(input).toHaveValue(10_000);
  });
});
