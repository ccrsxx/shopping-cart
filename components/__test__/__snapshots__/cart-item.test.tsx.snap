// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test CartItem renders correctly 1`] = `
<div>
  <li
    data-projection-id="1"
    style="opacity: 0; transform: translateX(50px) translateZ(0);"
  >
    <div
      class="flex rounded-lg border border-border-primary"
    >
      <a
        class="tab hidden rounded-r-none transition hover:brightness-90 focus-visible:brightness-90 md:block"
        href="/product/24"
      >
        <div
          class="relative h-[108px] w-[108px] rounded-l-lg shrink-0 shrink-0 bg-white animate-pulse !bg-primary"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="Emilia"
              class="!p-4"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
            <noscript />
          </span>
        </div>
      </a>
      <div
        class="flex w-full flex-col justify-between py-2 px-4"
      >
        <div>
          <a
            class="tab overflow-hidden text-ellipsis font-medium text-white/90 transition [display:-webkit-box] [-webkit-line-clamp:1] [-webkit-box-orient:vertical] hover:text-white"
            href="/product/24"
          >
            Emilia
          </a>
          <p
            class="font-bold"
          >
            $10.00
          </p>
        </div>
        <div
          class="flex flex-wrap justify-between text-sm font-light"
        >
          <div
            class="flex gap-1"
          >
            <p>
              Quantity: 
            </p>
            <input
              class="max-w-[60px] rounded-lg bg-background px-1 transition hover:ring-2 hover:ring-accent focus:outline-none focus:ring-2 focus:ring-accent"
              max="10000"
              min="1"
              type="number"
              value="1"
            />
          </div>
          <p>
            Total:
             
            <span
              class="font-semibold"
            >
              $10.00
            </span>
          </p>
        </div>
      </div>
      <div
        class="grid border-l border-border-primary text-sm inner:rounded-none"
      >
        <button
          aria-label="Increase"
          class="tab flex select-none items-center justify-center gap-2 rounded-lg bg-transparent py-2 px-3 
         font-medium transition duration-300 hover:text-primary active:scale-90 active:duration-150 
         enabled:hover:bg-accent disabled:cursor-not-allowed disabled:brightness-50 !rounded-tr-lg disabled:text-gray-600 disabled:!brightness-100"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </button>
        <button
          aria-label="Decrease"
          class="tab flex select-none items-center justify-center gap-2 rounded-lg bg-transparent py-2 px-3 
         font-medium transition duration-300 hover:text-primary active:scale-90 active:duration-150 
         enabled:hover:bg-accent disabled:cursor-not-allowed disabled:brightness-50 border-y border-border-primary disabled:text-gray-600 disabled:!brightness-100"
          disabled=""
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 13H5v-2h14v2z"
            />
          </svg>
        </button>
        <button
          aria-label="Delete"
          class="tab flex select-none items-center justify-center gap-2 rounded-lg bg-transparent py-2 px-3 
         font-medium transition duration-300 hover:text-primary active:scale-90 active:duration-150 
         enabled:hover:bg-accent disabled:cursor-not-allowed disabled:brightness-50 !rounded-br-lg"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </button>
      </div>
    </div>
  </li>
</div>
`;
